### PR TITLE
Rectangle/Ellipse rendering problems failing unit test/fix

### LINKED ISSUE #2282: Rectangle/Ellipse are not rendered when used in Checkbox template
Not all Rectangle/Ellipse are rendered properly when used in Checkbox template.

Repro steps:

1. Click CheckBox 'Check All'
2. All 10 listboxitems with checkboxes are expected to be checked, but only the first 3 are (visual/renderer problem).
3. 3 checkboxes at a time are checked, so  if you repeat 3 more time check all all checkboxes will be visually checked

Note: if we use default checkbox style all 10 will be checked, thus issue should be related to the Rectangle/Ellipse Shape (same checkbox template with ellipse shape have the same problem)

```XAML
            <StackPanel>
                <StackPanel.Styles>
                    <Style Selector="CheckBox.custom">
                        <Setter Property="Template">
                            <ControlTemplate>
                                <Grid ColumnDefinitions="Auto,*">
                                    <Border Name="border"
                                            Background="{TemplateBinding Background}"
                                            BorderBrush="{TemplateBinding BorderBrush}"
                                            BorderThickness="{TemplateBinding BorderThickness}"
                                            Width="18"
                                            Height="18"
                                            VerticalAlignment="Center">
                                        <Panel>
                                            <Rectangle Name="checkMark"
                                                  Fill="{DynamicResource HighlightBrush}"
                                                  Width="11"
                                                  Height="10"
                                                  Stretch="Uniform"
                                                  HorizontalAlignment="Center"
                                                  VerticalAlignment="Center"
                                    />
                                            <Rectangle Name="indeterminateMark"
                                                       Fill="{DynamicResource HighlightBrush}"
                                                       Width="10"
                                                       Height="10"
                                                       Stretch="Uniform"
                                                       HorizontalAlignment="Center"
                                                       VerticalAlignment="Center"/>
                                        </Panel>
                                    </Border>
                                    <ContentPresenter Name="PART_ContentPresenter"
                                                      TextBlock.Foreground="{TemplateBinding Foreground}"
                                                      ContentTemplate="{TemplateBinding ContentTemplate}"
                                                      Content="{TemplateBinding Content}"
                                                      Margin="4,0,0,0"
                                                      VerticalAlignment="{TemplateBinding VerticalContentAlignment}"
                                                      HorizontalAlignment="{TemplateBinding HorizontalContentAlignment}"
                                                      Grid.Column="1"/>
                                </Grid>
                            </ControlTemplate>
                        </Setter>
                    </Style>
                    <Style Selector="CheckBox.custom /template/ Rectangle#checkMark">
                        <Setter Property="IsVisible" Value="False"/>
                    </Style>
                    <Style Selector="CheckBox.custom:checked /template/ Rectangle#checkMark">
                        <Setter Property="IsVisible" Value="True"/>
                    </Style>
                </StackPanel.Styles>

                <CheckBox Name="selAll"  Content="Select All" />

                <ListBox Name="lb" Tag="{Binding #selAll.IsChecked}">
                    <ListBox.ItemTemplate>
                        <DataTemplate>
                            <CheckBox Classes="custom" Content="{Binding}" IsChecked="{Binding $parent[ListBox].Tag}" />
                        </DataTemplate>
                    </ListBox.ItemTemplate>
                </ListBox>
            </StackPanel>
```
code behind:
```C#
this.Get<ListBox>("lb").Items = Enumerable.Range(1, 10).Select(v => $"Item {v}").ToArray();
```

### PR DESCRIPTION
**Pull request by donandren**
## What does the pull request do?

failing unit test/fix for #2282

## What is the updated/expected behavior with this PR?

check issue #2282 details

## Checklist

- [x] Added unit tests (if possible)?

## Fixed issues
-fixes #2282

### CODE REVIEW

File: tests/Avalonia.Layout.UnitTests/LayoutManagerTests.cs [hunk 1]
```diff
@@ -238,7 +237,63 @@ public void Invalidating_Child_Remeasures_Parent()
             border.Height = 100;
 
             root.LayoutManager.ExecuteLayoutPass();
-            Assert.Equal(new Size(100, 100), panel.DesiredSize);             
+            Assert.Equal(new Size(100, 100), panel.DesiredSize);
+        }
+
+        [Fact]
+        public void Multiple_Shapes_When_Visibled_Should_Be_Arranged_Properly()
```
Comment (grokys) [inline tests/Avalonia.Layout.UnitTests/LayoutManagerTests.cs]: typo: should be "When_Visible`

File: src/Avalonia.Layout/LayoutManager.cs [hunk 2]
```diff
@@ -133,7 +133,7 @@ private void ExecuteMeasurePass()
 
         private void ExecuteArrangePass()
         {
-            while (_toArrange.Count > 0 && _toMeasure.Count == 0)
+            while (_toArrange.Count > 0)
```
Comment (grokys) [inline src/Avalonia.Layout/LayoutManager.cs]: Hmm, this doesn't feel like the correct solution. The idea is that when an arrange pass invalidates the measure of some control, the arrange pass is exited and the measure pass run again by [this loop](https://github.com/AvaloniaUI/Avalonia/blob/432795ab3aa34d26360cdbaee211069b837b4aec/src/Avalonia.Layout/LayoutManager.cs#L85).

Are you sure this is the correct fix?

**Review by grokys — CHANGES_REQUESTED** [review]

File: src/Avalonia.Layout/LayoutManager.cs (hunk 2 above)
Comment (donandren) [reply src/Avalonia.Layout/LayoutManager.cs]: Not sure this is the correct fix, but not sure how to improve it if needed.
But there is a problem with MaxPasses in [this loop](https://github.com/AvaloniaUI/Avalonia/blob/432795ab3aa34d26360cdbaee211069b837b4aec/src/Avalonia.Layout/LayoutManager.cs#L85) as they are final count (in our case 3).

So if we have more than 3 visuals than need remeasure in the arrange phase, we have a problem - layout arrange cycle will not be completed for all other visuals in the list (as illustrated in the unit test).

File: tests/Avalonia.Layout.UnitTests/LayoutManagerTests.cs (hunk 1 above)
Comment (donandren) [reply tests/Avalonia.Layout.UnitTests/LayoutManagerTests.cs]: what about 'When_Made_Visible' as the idea is: visual is non visible, then when it's made visible layout should performed properly?

Comment (grokys) [reply tests/Avalonia.Layout.UnitTests/LayoutManagerTests.cs]: Yes "When_Made_Visible" works 👍

File: src/Avalonia.Layout/LayoutManager.cs (hunk 2 above)
Comment (grokys) [reply src/Avalonia.Layout/LayoutManager.cs]: Hmm yeah. I'm going to take a deeper look into this.

Comment (donandren) [reply src/Avalonia.Layout/LayoutManager.cs]: @grokys  can you take a look at #2297 i think i've solved the issues with layoutmanager.
Changes in LayoutManager from #2297 should solve the problems also for this use case

**Comment by grokys** [comment]
Superseded by #2297

**Pull request closed without merging**